### ISSUE TITLE
clang-tidy does not always warn of too many arguments in call function

### ISSUE THREAD

**Issue by omarandlorraine** [llvm/llvm-project#60764]
Contents of `noarg.h`:

```h
#include <stdio.h>
int fn();
```

Contents of `noarg.c`:
```c
#include "noarg.h"

int fn() {
    return printf("hello");
}
```

Contents of `main.c`:
```c
#include "noarg.h"
int main() {
    fn(354);
}
```

Here we're calling `fn` with the wrong number of arguments. It's technically legal C (at least, before C23), but I feel that `clang-tidy` should warn us of the fact that declaring a function in this way: `int fn();` is bug-prone. 

It is worth noting that if Clang-Tidy sees the function definition *before* the call-site (not after), then we get a suitable warning. So, if we do this:

```shell
cat noarg.h noarg.c main.c > okay.c
clang-tidy okay.c
```

Then Clang-Tidy will warn us:

```console
/home/sam/tmp/okay.c:10:11: warning: too many arguments in call to 'fn' [clang-diagnostic-warning]
    fn(354);
          ^
```

which is correct and helpful. But Clang-Tidy does not warn us if we reverse the order of call-site and function:

```shell
cat noarg.h main.c noarg.c > not_okay.c
clang-tidy not_okay.c
```

we don't get the warning. Hopefully others can reproduce the issue.

Have I found a problem with Clang-Tidy here?

**Comment by PiotrZSL** [comment]
This is not Clang-Tidy issue.

in Clang with -std=c2x both result in:
main.c:6:8: error: too many arguments to function call, expected 0, have 1

In older standards you can enable following warnings:
```
main.c:6:7: warning: passing arguments to 'fn' without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
    fn(354);
      ^
main.c:9:7: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
int fn() {
      ^
       void
main.c:9:5: warning: no previous prototype for function 'fn' [-Wmissing-prototypes]
int fn() {
    ^
main.c:3:5: note: this declaration is not a prototype; add 'void' to make it a prototype for a zero-parameter function
int fn();
    ^
       void
```

**Comment by omarandlorraine** [comment]
Maybe the problem is not a clang-tidy specific thing. But do you agree it's the kind of sloppiness we expect linters to pull us up on?

Maybe the warnings you mention could be added to `-Wextra`, or a new grouping, `-Wextra2` or some such

**Comment by AaronBallman** [comment]
I don't think this behavior is a bug (in clang-tidy or in Clang itself). This is the way K&R C functions work in C. Clang issues a diagnostic in cases where we think the program behavior will change in C2x:
```
void func();

int main(void) {
  func(12); // warning: passing arguments to 'func' without a prototype is deprecated in all versions of C and is not supported in C2x
}
```
and clang-tidy faithfully reproduces that diagnostic: https://godbolt.org/z/oxE3bKWqa.

The problem with having clang-tidy claim it's a bugprone pattern is that there's a lot of existing C code that still uses this functionality (for example, in POSIX, this is how the `open` API is implemented -- https://man7.org/linux/man-pages/man2/open.2.html) and the uses aren't always prone to bugs. It boils down to programmer intent -- if the author of the API expected `()` to mean "accepts no arguments" as in C++ and C2x, then it's bugprone. If the author of the API expected `()` to mean "accepts zero or more arguments" as in K&R C, then it's behaving by design.

**Comment by AaronBallman** [comment]
Closing this as not a bug because it's behaving by design. However, if you want to continue the discussion about adding a new check to clang-tidy for K&R C functionality, please reopen the report with more details about how you envision the feature working (also: remove the `invalid` tag and add the `enhancement` tag instead).
